### PR TITLE
DQM plots for elastic scattering in 900 GeV run

### PR DESCRIPTION
**Pull request by jan-kaspar**
This PR introduces new DQM plots essential for monitoring of elastic scattering in the 900 GeV run foreseen at the end of November. Since elastic scattering is used for CTPPS calibration (alignment, optics), many of the new plots are of general interest for CTPPS.

We would like to test the code in online DQM on 8 Nov 2017 when the machine plans a test run at 900 GeV. Consequently a back-port of this PR to CMSSW_9_2_X will be prepared. Some code refinements may occur after the test run.

The code has been prepared such that the new plots can be easily disabled - by removing this python config line:
  https://github.com/jan-kaspar/cmssw/blob/dqm_rp_elastics/DQM/CTPPS/python/ctppsDQM_cff.py#L19

This PR can be tested with
  https://github.com/jan-kaspar/cmssw/blob/dqm_rp_elastics/DQM/CTPPS/test/elastic_dqm_test_cfg.py
which uses data from a CTPPS alignment run (July 2017) where also vertical RPs were inserted. A similar Roman Pot configuration is expected for the 900 GeV data.

Beyond the new DQM modules, this PR makes a minor change to the RP geometry: verticals in unit 220-fr are placed in a similar distance from beam as other vertical RPs.

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

**Comment by jan-kaspar** [comment]
b02a3bd implements a suggestion by Tomas Hreus how to enable/disable the online creation of the new plots without issuing further PRs.

**Comment by davidlange6** [comment]
please test

**Comment by dmitrijus** [comment]
+1

**Comment by ianna** [comment]
+1

File: DQM/CTPPS/plugins/ElasticPlotDQMSource.cc [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+//----------------------------------------------------------------------------------------------------
+
+ElasticPlotDQMSource::DiagonalPlots::DiagonalPlots(DQMStore::IBooker &ibooker, int id)
+{
+  // determine captions
+  bool top45 = id & 2;
+  bool top56 = id & 1;
+  bool diag = (top45 != top56);
+
+  char name[50];
```
Comment (davidlange6) [inline DQM/CTPPS/plugins/ElasticPlotDQMSource.cc]: hi @jan-kaspar - please use c++ style strings here
std::string name = ( (diag)? "diagonal" : "antidiagonal" ) + ...

File: DQM/CTPPS/plugins/ElasticPlotDQMSource.cc [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+
+    edm::EDGetTokenT< edm::DetSetVector<TotemRPRecHit> > tokenRecHit;
+    edm::EDGetTokenT< edm::DetSetVector<TotemRPUVPattern> > tokenUVPattern;
+    edm::EDGetTokenT< edm::DetSetVector<TotemRPLocalTrack> > tokenLocalTrack;
+
+    /// plots related to one (anti)diagonal
+    struct DiagonalPlots
+    {
+      // in the order 45-220-fr, 45-210-fr, 56-210-fr, 56-220-fr
+      std::vector<unsigned int> rpIds;
```
Comment (davidlange6) [inline DQM/CTPPS/plugins/ElasticPlotDQMSource.cc]: change to a std::array as it has a fixed length of 4

File: DQM/CTPPS/plugins/ElasticPlotDQMSource.cc [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+
+      // track correlation in vertical RPs
+      MonitorElement *h2_track_corr_vert=nullptr;
+
+      // y distributions in a RP (1st index) with another RP (2nd index) in coincidence
+      std::vector<std::vector<MonitorElement *>> v_h_y;
```
Comment (davidlange6) [inline DQM/CTPPS/plugins/ElasticPlotDQMSource.cc]: similarly an array of arrays

File: DQM/CTPPS/plugins/ElasticPlotDQMSource.cc [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+    {
+      // loop over RPs
+      for (unsigned int rp = 0; rp < 6; ++rp)
+      {
+        // skip horizontals - irrelevant for elastics
+        if (rp == 2 || rp == 3)
+          continue;
+
+        // skip "nr" units - not equipped
+        if (rp <= 2)
```
Comment (davidlange6) [inline DQM/CTPPS/plugins/ElasticPlotDQMSource.cc]: you can start the loop at 3?

Comment (davidlange6) [reply DQM/CTPPS/plugins/ElasticPlotDQMSource.cc]: actually 4...

File: DQM/CTPPS/plugins/ElasticPlotDQMSource.cc [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+    unsigned int n_pat_u = 0, n_pat_v = 0;
+    for (auto &p : ds)
+    {
+      if (! p.getFittable())
+        continue;
+
+      if (p.getProjection() == TotemRPUVPattern::projU)
+        n_pat_u++;
+
+      if (p.getProjection() == TotemRPUVPattern::projV)
```
Comment (davidlange6) [inline DQM/CTPPS/plugins/ElasticPlotDQMSource.cc]: can be an else if?

File: DQM/CTPPS/plugins/ElasticPlotDQMSource.cc [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+    void analyze(edm::Event const& e, edm::EventSetup const& eSetup) override;
+
+    void beginLuminosityBlock(edm::LuminosityBlock const& lumi, edm::EventSetup const& eSetup) override;
+
+    void endLuminosityBlock(edm::LuminosityBlock const& lumi, edm::EventSetup const& eSetup) override;
+
+    void endRun(edm::Run const& run, edm::EventSetup const& eSetup) override;
+
+  private:
+    static constexpr double ls_duration = 23.357; // s
```
Comment (davidlange6) [inline DQM/CTPPS/plugins/ElasticPlotDQMSource.cc]: better to cache 1/ls_duration, which seems to be what is actually used.

**Comment by jan-kaspar** [comment]
Hi @davidlange6, thanks! I will address your comments after digesting the machine test today - it may come out that some more changes are needed.

**Comment by davidlange6** [comment]
-1
pending answers to comments

**Comment by fabiocos** [comment]
@jan-kaspar could you please come back to this PR? Is this still needed?

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

**Comment by jan-kaspar** [comment]
Hi all,

Sorry for the late reply. This PR is for a special low energy run. The run
was postponed and so far not rescheduled. As the PR has become unmergeable,
I propose it is closed now and I will come back when relevant (possibly
only in Run III). Sorry for the lack of clarification.

Best wishes,

Jan

**Comment by fabiocos** [comment]
Unless you have some short term use case for this code, yes please close it.

**Pull request closed without merging**